packages/unified-selection/src/unified-selection/SelectionScope.ts

```diff
@@ -0,0 +1,279 @@
+                    )
+                    SELECT DISTINCT ECInstanceId, ec_classname(ECClassId, 's.c') AS ClassName FROM Elements
+                    WHERE ${recurseUntilRoot ? "" : "Depth = 0 OR"} ParentId IS NULL;`;
+
+    yield* executeQuery(queryExecutor, query, bindings);
+  }
+
+  async function* computeCategorySelection(queryExecutor: IECSqlQueryExecutor, ids: string[]): AsyncIterableIterator<SelectableInstanceKey> {
+    const bindings: ECSqlBinding[] = [];
+    const query = `SELECT DISTINCT c.ECInstanceId, ec_classname(c.ECClassId, 's.c') AS ClassName
+                   FROM BisCore.Category c JOIN BisCore.GeometricElement2d g ON g.Category.Id = c.ECInstanceId
+                   WHERE ${formIdBindings("g.ECInstanceId", ids, bindings)}
+                   UNION ALL
+                   SELECT DISTINCT c.ECInstanceId, ec_classname(c.ECClassId, 's.c') AS ClassName
+                   FROM BisCore.Category c JOIN BisCore.GeometricElement3d g ON g.Category.Id = c.ECInstanceId
+                   WHERE ${formIdBindings("g.ECInstanceId", ids, bindings)};`;
+
+    yield* executeQuery(queryExecutor, query, bindings);
+  }
+
+  async function* computeModelSelection(queryExecutor: IECSqlQueryExecutor, ids: string[]): AsyncIterableIterator<SelectableInstanceKey> {
+    const bindings: ECSqlBinding[] = [];
+    const query = `SELECT DISTINCT m.ECInstanceId, ec_classname(m.ECClassId, 's.c') AS ClassName
+                   FROM BisCore.Model m JOIN BisCore.Element e ON e.Model.Id = m.ECInstanceId
+                   WHERE ${formIdBindings("e.ECInstanceId", ids, bindings)};`;
+
+    yield* executeQuery(queryExecutor, query, bindings);
+  }
+
+  async function* computeFunctionalElementSelection(queryExecutor: IECSqlQueryExecutor, ids: string[]): AsyncIterableIterator<SelectableInstanceKey> {
+    const bindings: ECSqlBinding[] = [];
+    const query = `WITH RECURSIVE
+                    ${getNearestFunctionalElementQuery(ids, bindings)}
+                    Element3dRelatedFunctionalElement (ECInstanceId, ECClassId, FunctionalECInstanceId, FunctionalECClassId) AS (
+                      SELECT ge.ECInstanceId, ge.ECClassId, r.TargetECInstanceId, r.TargetECClassId FROM BisCore.GeometricElement3d ge
+                        LEFT JOIN Functional.PhysicalElementFulfillsFunction r ON r.SourceECInstanceId = ge.ECInstanceId
+                      WHERE ${formIdBindings("ge.ECInstanceId", ids, bindings)}
+                    )
+                    SELECT ECInstanceId, ec_classname(ECClassId, 's.c') AS ClassName FROM MergedElements
+                    UNION ALL
+                    SELECT ECInstanceId, ec_classname(ECClassId, 's.c') AS ClassName FROM Element3dRelatedFunctionalElement WHERE FunctionalECInstanceId IS NULL
+                    UNION ALL
+                    SELECT FunctionalECInstanceId AS ECInstanceId, ec_classname(FunctionalECClassId, 's.c') AS ClassName FROM Element3dRelatedFunctionalElement WHERE FunctionalECInstanceId IS NOT NULL;`;
```

Comment:
could we instead do this?

```sql
SELECT 
  COALESCE(TargetECInstanceId, SourceECInstanceId) AS ECInstanceId,
  COALESCE(TargetECClassId, SourceECClassId) AS ECClassId
FROM Functional.PhysicalElementFulfillsFunction
WHERE ${formIdBindings("SourceECInstanceId", ids, bindings)}
```

Reply:
This would not return the original elements that do not have an associated `PhysicalElementFulfillsFunction` relationship, however with some small additional tweaks I was able to use this approach.